examples/gno.land/r/gov/dao/v3/impl/types.gno

```diff
@@ -0,0 +1,198 @@
+	return ProposalsStatuses{avl.NewTree()}
+}
+
+func (pss ProposalsStatuses) GetStatus(id dao.ProposalID) *proposalStatus {
+	pids := ufmt.Sprintf("%v", int(id))
+	psv, ok := pss.Get(pids)
+	if !ok {
+		return nil
+	}
+
+	ps, ok := psv.(*proposalStatus)
+	if !ok {
+		panic("ProposalsStatuses must contains only proposalStatus types")
+	}
+
+	return ps
+}
+
+type proposalStatus struct {
+	YesVotes memberstore.MembersByTier
+	NoVotes  memberstore.MembersByTier
+	AllVotes memberstore.MembersByTier
+
+	Accepted bool
+	Denied   bool
+
+	DeniedReason string
+
+	TiersAllowedToVote []string
+
+	TotalPower float64 // TotalPower is the power of all the members existing when this proposal was created.
+}
+
+func newProposalStatus(allowedToVote []string) *proposalStatus {
+	return &proposalStatus{
+		YesVotes: memberstore.NewMembersByTier(),
+		NoVotes:  memberstore.NewMembersByTier(),
+		AllVotes: memberstore.NewMembersByTier(),
+
+		TiersAllowedToVote: allowedToVote,
+
+		TotalPower: memberstore.Get().GetTotalPower(),
+	}
+}
+
+func (ps *proposalStatus) YesPercent() float64 {
+	var yp float64
+
+	memberstore.Tiers.Iterate("", "", func(tn string, ti interface{}) bool {
+		tier, ok := ti.(memberstore.Tier)
+		if !ok {
+			panic("type must be memberstore.Tier")
+		}
+
+		power := tier.PowerHandler(memberstore.Get(), memberstore.Tiers)
+		ts := ps.YesVotes.GetTierSize(tn)
+
+		yp = yp + (power * float64(ts))
+
+		return false
+	})
+
+	return (yp / ps.TotalPower) * 100
+}
+
+func (ps *proposalStatus) NoPercent() float64 {
+	var np float64
+
+	memberstore.Tiers.Iterate("", "", func(tn string, ti interface{}) bool {
+		tier, ok := ti.(memberstore.Tier)
+		if !ok {
+			panic("type must be memberstore.Tier")
+		}
+
+		power := tier.PowerHandler(memberstore.Get(), memberstore.Tiers)
+		ts := ps.NoVotes.GetTierSize(tn)
+		np = np + (power * float64(ts))
+
+		return false
+	})
+
+	return (np / ps.TotalPower) * 100
+}
+
+func (ps *proposalStatus) IsAllowed(tier string) bool {
+	for _, ta := range ps.TiersAllowedToVote {
+		if ta == tier {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (ps *proposalStatus) String() string {
+	var sb strings.Builder
+	sb.WriteString("### Proposal Status:\n\n")
+
+	if ps.Accepted {
+		sb.WriteString("- **PROPOSAL HAS BEEN ACCEPTED**\n")
+	} else if ps.Denied {
+		sb.WriteString("- **PROPOSAL HAS BEEN DENIED**\n")
+		if ps.DeniedReason != "" {
+			sb.WriteString("REASON: ")
+			sb.WriteString(ps.DeniedReason)
+		}
+	} else {
+		sb.WriteString("- **Proposal open for votes**\n")
+	}
+
+	sb.WriteString("- Allowed tiers to vote: ")
+	for _, t := range ps.TiersAllowedToVote {
+		sb.WriteString(t)
+		sb.WriteString(" ")
+	}
+	sb.WriteString("\n")
+
+	sb.WriteString(ufmt.Sprintf("- YES PERCENT: %v%%\n", ps.YesPercent()))
+	sb.WriteString(ufmt.Sprintf("- NO PERCENT: %v%%\n", ps.NoPercent()))
+
+	return sb.String()
+}
+
+func StringifyVotes(ps *proposalStatus) string {
+	var sb strings.Builder
+
+	writeVotes(&sb, ps.YesVotes, "### Voting YES")
```

Comment:
```suggestion
	writeVotes(&sb, ps.YesVotes, "YES")
```

It seems more appropriate since after we're concatenating everything on a single line at the end. This increases cognitive complexity compared to simply specifying the word that changes and allowing "writeVotes" to handle the H3, etc.

Reply:
changed